### PR TITLE
#142 : [New Rule] No imports from static tests namespaces

### PR DESCRIPTION
**Pull request by konarshankar07**
New ruleset added which will prevent importing static test namespaces
Fixed issue : https://github.com/magento/magento-coding-standard/issues/142

### CODE REVIEW

File: Magento2/Sniffs/Namespaces/UseDeclarationSniff.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Detects static test namespace.
+ */
+class UseDeclarationSniff implements Sniff
+{
+
+    /**
+     * @var string
+     */
+    private $_prohibitNamespace = 'Magento\Tests';
```
Comment (lenaorobei) [inline Magento2/Sniffs/Namespaces/UseDeclarationSniff.php]:
```suggestion
    private $prohibitNamespace = 'Magento\Tests';
```

File: Magento2/Sniffs/Namespaces/UseDeclarationSniff.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * @var string
+     */
+    protected $warningMessage = 'Incorrect namespace has been imported.';
```
Comment (lenaorobei) [inline Magento2/Sniffs/Namespaces/UseDeclarationSniff.php]:
```suggestion
    protected $warningMessage = 'Application modules should not use classed from test modules.';
```

File: Magento2/Sniffs/Namespaces/UseDeclarationSniff.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * @var string
+     */
+    protected $warningCode = 'WrongImportNamespaces';
```
Comment (lenaorobei) [inline Magento2/Sniffs/Namespaces/UseDeclarationSniff.php]:
```suggestion
    protected $warningCode = 'WrongImport';
```

File: Magento2/Tests/Namespaces/UseDeclarationUnitTest.inc [hunk 4]
```diff
@@ -0,0 +1,4 @@
+<?php
+use Magento\Tests;
+use Magento\Foo\Tests as FakeTest;
+use Magento\Real\Classes;
```
Comment (lenaorobei) [inline Magento2/Tests/Namespaces/UseDeclarationUnitTest.inc]: Please add cases (positive and negative) with group use declarations.

File: Magento2/Sniffs/Namespaces/UseDeclarationSniff.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+ */
+namespace Magento2\Sniffs\Namespaces;
+
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHP_CodeSniffer\Files\File;
+
+/**
+ * Detects static test namespace.
+ */
+class UseDeclarationSniff implements Sniff
```
Comment (lenaorobei) [inline Magento2/Sniffs/Namespaces/UseDeclarationSniff.php]:
```suggestion
class ImportsFromTestNamespaceSniff implements Sniff
```

File: Magento2/ruleset.xml [hunk 6]
```diff
@@ -189,6 +189,10 @@
         <exclude-pattern>*Test.php</exclude-pattern>
         <exclude-pattern>*/tests/*</exclude-pattern>
     </rule>
+    <rule ref="Magento2.Namespaces.UseDeclaration">
```
Comment (lenaorobei) [inline Magento2/ruleset.xml]: Please add exclude for test files.

File: Magento2/Sniffs/Namespaces/UseDeclarationSniff.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
+<?php
```
Comment (ihor-sviziev) [inline Magento2/Sniffs/Namespaces/UseDeclarationSniff.php]: I'd like to have `UseDeclarationSniff.md` file that will describe the issue behind

**Comment by konarshankar07** [comment]
I'm still working on group namespace changes

File: Magento2/ruleset.xml [hunk 8]
```diff
@@ -189,6 +189,10 @@
         <exclude-pattern>*Test.php</exclude-pattern>
         <exclude-pattern>*/tests/*</exclude-pattern>
     </rule>
+    <rule ref="Magento2.Namespaces.ImportsFromTestNamespace">
+        <severity>8</severity>
+        <type>warning</type>
```
Comment (ihor-sviziev) [inline Magento2/ruleset.xml]: Please exclude tests, like this:
```suggestion
        <type>warning</type>
        <exclude-pattern>*/_files/*</exclude-pattern>
        <exclude-pattern>*/Fixtures/*</exclude-pattern>
        <exclude-pattern>*/Test/*</exclude-pattern>
        <exclude-pattern>*Test.php</exclude-pattern>
        <exclude-pattern>*/tests/*</exclude-pattern>
```

Comment (ihor-sviziev) [reply Magento2/ruleset.xml]: Actually it's so minor, I added it myself

Comment (konarshankar07) [reply Magento2/ruleset.xml]: Thanks @ihor-sviziev

**Comment by ihor-sviziev** [comment]
@lenaorobei could you review this PR?

**Comment by lenaorobei** [comment]
Changes look good to me. I'll run against Magento code to see if there is no issues and will merge it.

**Pull request merged**